plugins/cxaudio/fu-cxaudio-device.c

```diff
@@ -0,0 +1,889 @@
+		guint8 value = 0;
+		if (!fu_cxaudio_device_EepromWrite (self,
+						    CXDEV_EEPROM_LAYOUT_SIGNATURE_ADDRESS,
+						    &value, sizeof(value), error)) {
+			g_prefix_error (error, "failed to initialize layout signature ");
+			return FALSE;
+		}
+		if (!fu_cxaudio_device_EepromWrite (self,
+						    CXDEV_EEPROM_LAYOUT_VERSION_ADDRESS,
+						    &value, sizeof(value), error)) {
+			g_prefix_error (error, "failed to initialize layout signature ");
+			return FALSE;
+		}
+		g_debug ("initialized layout signature");
+	}
+#if 0
+	gboolean bShouldProtectSerialNumber =
+		self->eeprom_layout_version >= 1 &&
+		fu_cxaudio_firmware_get_layout_version (FU_CXAUDIO_FIRMWARE (firmware)) >= 0 &&
+		self->serial_number_set &&
+		fu_cxaudio_firmware_get_vid (FU_CXAUDIO_FIRMWARE (firmware)) == fu_usb_device_get_vid (FU_USB_DEVICE (device));
+	g_debug ("EepromPatch::shouldProtectSerialNumber: 0x%x", (guint32)bShouldProtectSerialNumber);
+#endif
+
+	/* perform the actual write */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	for (guint i = 0; i < records->len; i++) {
+		FuSrecFirmwareRecord *rcd = g_ptr_array_index (records, i);
+		if (rcd->kind != FU_FIRMWARE_SREC_RECORD_KIND_S3_DATA_32)
+			continue;
+		g_debug ("writing @0x%04x len:0x%02x", rcd->addr, rcd->buf->len);
+		if (!fu_cxaudio_device_MemoryOperation (self,
+							CXDEV_OPERATION_WRITE,
+							CXDEV_MEMTYPE_EEPROM,
+							rcd->addr,
+							rcd->buf->data, rcd->buf->len,
+							OPT_VERIFY_MEM_WRITE,
+							error)) {
+			g_prefix_error (error, "failed to write @0x%04x len:0x%02x: ",
+					rcd->addr, rcd->buf->len);
+			return FALSE;
+		}
+		fu_device_set_progress_full (device, (gsize) i, (gsize) records->len);
+	}
+
+	/* in case of a full FW upgrade invalidate the old FW patch (if any)
+	 * as it may have not been done by the S37 file */
+	if (FileType == CXDEV_FILETYPE_CX2070X_FW) {
+		CXDEV_EEPROM_CX2070X_PATCH_INFO pinfo = { 0 };
+		if (!fu_cxaudio_device_EepromRead (self,
+						   CXDEV_EEPROM_CX2070X_PATCH_INFO_OFFSET,
+						   (guint8 *) &pinfo,
+						   sizeof(pinfo),
+						   error)) {
+			g_prefix_error (error, "failed to read EEPROM patch info: ");
+			return FALSE;
+		}
+		if (pinfo.PatchSignature == SIGNATURE_BYTE_CX2070X_PATCH) {
+			memset (&pinfo, 0x0, sizeof(pinfo));
+			if (!fu_cxaudio_device_EepromWrite (self,
+							    CXDEV_EEPROM_CX2070X_PATCH_INFO_OFFSET,
+							    (guint8 *) &pinfo,
+							    sizeof(pinfo),
+							    error)) {
+				g_prefix_error (error, "failed to write empty EEPROM patch info");
+				return FALSE;
+			}
+			g_debug ("invalidated old FW patch for CX2070x (RAM) device");
+		}
+	}
+
+	/* unpark the FW */
+	if (!fu_cxaudio_device_RegisterClearBit (self, 0x1000, 7, error))
+		return FALSE;
+
+	/* if supported, self reset */
+	if (self->sw_reset_supported) {
+		fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+		return fu_cxaudio_device_reset (self, error);
```

Comment:
What do you do if it's not supported?  Need to power cycle the device?  Or is that only an EVB type thing that it's controlled by jumper?

Reply:
EBV and old devices we don't care about. Power cycling an audio chip might be unplug-replug or even a cold boot if it's on a motherboard.